rebac-admin-backend/v1/validation.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2024 Canonical Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/canonical/identity-platform-admin-ui/rebac-admin-backend/v1/resources"
+	"github.com/go-playground/validator/v10"
+)
+
+// handlerWithValidation decorates a given handler with validation logic. The
+// request body is parsed into a safely-typed value and passed to the handler
+// via context.
+type handlerWithValidation struct {
+	// TODO(babakks) remove the embedded struct.
+	resources.Unimplemented
+
+	handler resources.ServerInterface
+}
+
+// newHandlerWithValidation returns a new instance of the validationHandlerDecorator struct.
+func newHandlerWithValidation(handler resources.ServerInterface) *handlerWithValidation {
+	return &handlerWithValidation{
+		handler: handler,
+	}
+}
+
+// requestBodyContextKey is the context key to retrieve the parsed request body struct instance.
+type requestBodyContextKey struct{}
+
+// getRequestBodyFromContext fetches request body from given context.
+func getRequestBodyFromContext[T any](ctx context.Context) (*T, error) {
+	validator.New()
+	if body, ok := ctx.Value(requestBodyContextKey{}).(*T); ok {
+		return body, nil
+	}
+	return nil, NewMissingRequestBodyError("request body is not available")
+}
+
+// newRequestWithBodyInContext sets the given body in a new request instance context
+// and returns the new request.
+//
+// Note that, technically, this method could be an ordinary (non-generic) method,
+// but it's defined as one to avoid confusion over value vs pointer arguments.
+func newRequestWithBodyInContext[T any](r *http.Request, body *T) *http.Request {
+	return r.WithContext(context.WithValue(r.Context(), requestBodyContextKey{}, body))
+}
+
+// parseRequestBody parses request body as JSON and populates the given body instance.
+func parseRequestBody[T any](r *http.Request) (*T, error) {
+	body := new(T)
+	defer r.Body.Close()
+	if err := json.NewDecoder(r.Body).Decode(body); err != nil {
+		return nil, NewMissingRequestBodyError("request body is not a valid JSON")
+	}
+	return body, nil
+}
+
+// setRequestBodyInContext is a helper method to avoid repetition. It parses
+// request body and if it's okay, will delegate to the provided callback with a
+// new HTTP request instance with the parse body in the context.
+func setRequestBodyInContext[T any](w http.ResponseWriter, r *http.Request, f func(w http.ResponseWriter, r *http.Request, body *T)) {
+	body, err := parseRequestBody[T](r)
+	if err != nil {
+		writeErrorResponse(w, err)
+		return
+	}
+	f(w, newRequestWithBodyInContext(r, body), body)
+}
```

Comment:
this looks really smart, but i m finding it really hard to understand

the use of generics could be dropped if u offload [this](https://github.com/canonical/identity-platform-admin-ui/pull/234/files#diff-17dd23c5e4825273ddfc3f074394963daa15bc9a48de3e41d1d9433e6932e7f7R54) to each implementation in [here]https://github.com/canonical/identity-platform-admin-ui/pull/234/files#diff-1088fa4484e7da34fc7f69eee616d0f5ad686a02941cd9f6c0b28d737c1a8cbeR19]

while for `validation_utils.go` generics make a lot of sense, here i m not 100% sure, especially if we read the sections in https://go.dev/blog/when-generics on `when not to use`

Reply:
I'm not sure if I'm getting your point correctly, but I don't have a strong opinion on using generics for this part. Of course, parsing/validation step can be done without generics. For example, without generics the handler methods in the `handlerWithValidation` struct would look change from this:

```go
func (v handlerWithValidation) PostGroups(w http.ResponseWriter, r *http.Request) {
  setRequestBodyInContext[resources.Group](w, r, func(w http.ResponseWriter, r *http.Request, body *resources.Group) {
    if err := validateGroup(body); err != nil {
      writeErrorResponse(w, err)
      return
    }
    v.handler.PostGroups(w, r)
  })
}
```

to this:

```go
func (v handlerWithValidation) PostGroups(w http.ResponseWriter, r *http.Request) {
  body := &resources.Group{}
  setRequestBodyInContext(body, w, r, func(w http.ResponseWriter, r *http.Request) {
    if err := validateGroup(body); err != nil {
      writeErrorResponse(w, err)
      return
    }
    v.handler.PostGroups(w, r)
  })
}
```

Is this what you're asking for?

Reply:
the 2 bits of code above are basically the same, if we can achieve something without generics we should

it arguably makes the code harder to read, it doesn't speed it up and doesn't really bring anything to the plate apart from sugar syntax in this case

Reply:
@alesstimec Yeah, I've seen generics used just for ensuring type safety and helping with type casting in other legit OSS projects (other languages).